taurus/util/tests/test_flatten.py

```diff
@@ -0,0 +1,54 @@
+from taurus.entity.bounds.categorical_bounds import CategoricalBounds
+from taurus.entity.object import ProcessSpec, MaterialSpec, IngredientSpec, ProcessRun, \
+    MaterialRun, IngredientRun
+from taurus.entity.template.condition_template import ConditionTemplate
+from taurus.entity.template.process_template import ProcessTemplate
+from taurus.util import flatten, recursive_flatmap
+
+
+def test_flatten_bounds():
+    """Test that flatten works when the objects contain other objects."""
+    bounds = CategoricalBounds(categories=["foo", "bar"])
+    template = ProcessTemplate(
+        "spam",
+        conditions=[(ConditionTemplate(name="eggs", bounds=bounds), bounds)]
+    )
+    spec = ProcessSpec(name="spec", template=template)
+
+    flat = flatten(spec)
+    assert len(flat) == 2, "Expected 2 flattened objects"
+
+
+def test_flatten_empty_history():
+    """Test that flatten works when the objects are empty and go through a whole history."""
+    procured = ProcessSpec(name="procured")
+    input = MaterialSpec(name="foo", process=procured)
+    transform = ProcessSpec(name="transformed")
+    ingredient = IngredientSpec(name="input", material=input, process=transform)
+
+    procured_run = ProcessRun(name="procured", spec=procured)
+    input_run = MaterialRun(name="foo", process=procured_run, spec=input)
+    transform_run = ProcessRun(name="transformed", spec=transform)
+    ingredient_run = IngredientRun(
```

Comment:
`name` parameter is vestigial for IngredientRuns

```suggestion
    ingredient_run = IngredientRun(material=input_run, process=transform_run, spec=ingredient)
```